File: src/dependent/notifications/EmailService.java
package org.example.dependent.notifications;

import java.util.HashMap;
import java.util.Map;

public class EmailService {
    private final TemplateRenderer templateRenderer;
    private final SmtpClient smtpClient;

    public EmailService(TemplateRenderer templateRenderer, SmtpClient smtpClient) {
        this.templateRenderer = templateRenderer;
        this.smtpClient = smtpClient;
    }

    public String sendOrderConfirmation(String orderId, String toEmail) {
        Map<String, Object> data = new HashMap<>();
        data.put("orderId", orderId);

        String subject = "Your order " + orderId + " is confirmed";
        String body = templateRenderer.render("order_confirmation", data);
        return smtpClient.send(toEmail, subject, body);
    }

    public String sendPasswordReset(String token, String toEmail) {
        Map<String, Object> data = new HashMap<>();
        data.put("token", token);

        String subject = "Password reset instructions";
        String body = templateRenderer.render("password_reset", data);
        return smtpClient.send(toEmail, subject, body);
    }
}

File: src/dependent/notifications/email_service.py
from typing import Any, Dict

from .template_renderer import TemplateRenderer
from .smtp_client import SmtpClient


class EmailService:
    def __init__(self, template_renderer: TemplateRenderer, smtp_client: SmtpClient) -> None:
        self._template_renderer = template_renderer
        self._smtp_client = smtp_client

    def send_order_confirmation(self, order_id: str, to_email: str) -> str:
        data: Dict[str, Any] = {"orderId": order_id}
        subject = f"Your order {order_id} is confirmed"
        body = self._template_renderer.render("order_confirmation", data)
        return self._smtp_client.send(to_email, subject, body)

    def send_password_reset(self, token: str, to_email: str) -> str:
        data: Dict[str, Any] = {"token": token}
        subject = "Password reset instructions"
        body = self._template_renderer.render("password_reset", data)
        return self._smtp_client.send(to_email, subject, body)
